src/core/search/ast_expr.cc

```diff
@@ -13,57 +13,33 @@ using namespace std;
 
 namespace dfly::search {
 
-AstTermNode::AstTermNode(std::string term)
-    : term_{move(term)}, pattern_{"\\b" + term_ + "\\b", std::regex::icase} {
+AstTermNode::AstTermNode(string term)
+    : term{term}, pattern{"\\b" + term + "\\b", std::regex::icase} {
 }
 
-bool AstTermNode::Check(SearchInput input) const {
-  return input.Check([this](string_view str) {
-    return regex_search(str.begin(), str.begin() + str.size(), pattern_);
-  });
+AstRangeNode::AstRangeNode(int64_t lo, int64_t hi) : lo{lo}, hi{hi} {
 }
 
-string AstTermNode::Debug() const {
-  return "term{" + term_ + "}";
+AstNegateNode::AstNegateNode(AstNode&& node) : node{make_unique<AstNode>(move(node))} {
 }
 
-bool AstNegateNode::Check(SearchInput input) const {
-  return !node_->Check(input);
-}
-
-string AstNegateNode::Debug() const {
-  return "not{" + node_->Debug() + "}";
-}
-
-bool AstLogicalNode::Check(SearchInput input) const {
-  return op_ == kOr ? (l_->Check(input) || r_->Check(input))
-                    : (l_->Check(input) && r_->Check(input));
-}
-
-string AstLogicalNode::Debug() const {
-  string op = op_ == kOr ? "or" : "and";
-  return op + "{" + l_->Debug() + "," + r_->Debug() + "}";
-}
-
-bool AstFieldNode::Check(SearchInput input) const {
-  return node_->Check(SearchInput{input, field_});
-}
-
-string AstFieldNode::Debug() const {
-  return "field:" + field_ + "{" + node_->Debug() + "}";
-}
+AstLogicalNode::AstLogicalNode(AstNode&& l, AstNode&& r, LogicOp op) : op{op}, nodes{} {
```

Comment:
I'm not super fan of doing simplifications like this in the constructor, but if we do it then at least construct first and put the flattening logic into a new `Simplify` function.

Reply:
Construct first, but in what state - I don't want to allocate the vector just to destruct it